lib/nethealth/nethealth.go

```diff
@@ -0,0 +1,587 @@
+	heartbeatInterval = 1 * time.Second
+
+	// resyncInterval is the duration between full resyncs of local state with kubernetes. If a node is deleted it
+	// may not be detected until the full resync completes.
+	resyncInterval = 15 * time.Minute
+
+	// dnsDiscoveryInterval is the duration of time for doing DNS based service discovery for pod changes. This is a
+	// lightweight test for whether there is a change to the nethealth pods within the cluster.
+	dnsDiscoveryInterval = 10 * time.Second
+
+	// Default selector to use for finding nethealth pods
+	DefaultSelector = "k8s-app=nethealth"
+
+	// DefaultServiceDiscoveryQuery is the default name to query for service discovery changes
+	DefaultServiceDiscoveryQuery = "any.nethealth"
+)
+
+const (
+	INIT    = "init"
+	UP      = "up"
+	TIMEOUT = "timeout"
+)
+
+type Config struct {
+	// PrometheusPort is the port to bind to for serving prometheus metrics
+	PrometheusPort uint32
+
+	// Namespace is the kubernetes namespace to monitor for other nethealth instances
+	Namespace string
+	// NodeName is the node this instance is running on
+	NodeName string
+	// Selector is a kubernetes selector to find all the nethealth pods in the configured namespace
+	Selector string
+	// ServiceDiscoveryQuery is a DNS name that will be used for lightweight service discovery checks. A query to
+	// any.<service>.default.svc.cluster.local will return a list of pods for the service. If the list of pods
+	// changes we know to resync with the kubernetes API. This method uses significantly less resources than running a
+	// kubernetes watcher on the API. Defaults to any.nethealth which will utilize the search path from resolv.conf.
+	ServiceDiscoveryQuery string
+}
+
+func (c Config) New() (*Server, error) {
+
+	promPeerRTT := prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: "nethealth",
+		Subsystem: "echo",
+		Name:      "duration_seconds",
+		Help:      "The round trip time to reach the peer",
+		Buckets: []float64{
+			0.0001, // 0.1 ms
+			0.0002, // 0.2 ms
+			0.0003, // 0.3 ms
+			0.0004, // 0.4 ms
+			0.0005, // 0.5 ms
+			0.0006, // 0.6 ms
+			0.0007, // 0.7 ms
+			0.0008, // 0.8 ms
+			0.0009, // 0.9 ms
+			0.001,  // 1ms
+			0.0015, // 1.5ms
+			0.002,  // 2ms
+			0.003,  // 3ms
+			0.004,  // 4ms
+			0.005,  // 5ms
+			0.01,   // 10ms
+			0.02,   // 20ms
+			0.040,  // 40ms
+			0.080,  // 80ms
+		},
+	}, []string{"node_name", "peer_name"})
+	promPeerTimeout := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "nethealth",
+		Subsystem: "echo",
+		Name:      "timeout_total",
+		Help:      "The number of echo requests that have timed out",
+	}, []string{"node_name", "peer_name"})
+	promPeerRequest := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "nethealth",
+		Subsystem: "echo",
+		Name:      "request_total",
+		Help:      "The number of echo requests that have been sent",
+	}, []string{"node_name", "peer_name"})
+
+	prometheus.MustRegister(
```

Comment:
I believe MustRegister panics on failure. If you want to return and process the error there is a prometheus.Register function, which shouldn't be needed for this. Otherwise it requires placing in an init function, which is also considered a bad practice and will get noticed by the linters.